From 0b5f3e46c9ba376e37739ac3eb5b24ed95fb5f61 Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Thu, 23 Dec 2021 09:27:47 +0800
Subject: [PATCH 22/27] medai: aspeed: Fix using sync mode if refresh-rate
 changed only

If only refresh-rate is different in timings change, user could skip
set_dv_timings since the size of frame is the same. In this way,
capture mode will not be updated per frame-size and stay at default
, sync, mode.

To fix the problem, we just go aspeed_video_set_resolution() if
aspeed_video_resolution_work() finishes its job.

Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
Change-Id: Iaee4f42840011ffddba3b715510b97d29fdbaee6
(cherry picked from commit 05a2090ea5775ba419a657b0473c356b6e48b36c)
---
 drivers/media/platform/aspeed-video.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/aspeed-video.c b/drivers/media/platform/aspeed-video.c
index 600df3f17edd..41197cb294cc 100644
--- a/drivers/media/platform/aspeed-video.c
+++ b/drivers/media/platform/aspeed-video.c
@@ -1699,6 +1699,7 @@ static void aspeed_video_resolution_work(struct work_struct *work)
 	struct delayed_work *dwork = to_delayed_work(work);
 	struct aspeed_video *video = container_of(dwork, struct aspeed_video,
 						  res_work);
+	bool is_res_chg = false;
 
 	aspeed_video_on(video);
 
@@ -1712,8 +1713,15 @@ static void aspeed_video_resolution_work(struct work_struct *work)
 
 	aspeed_video_get_resolution(video);
 
-	if (video->detected_timings.width != video->active_timings.width ||
-	    video->detected_timings.height != video->active_timings.height) {
+	if (video->v4l2_input_status)
+		goto done;
+
+	is_res_chg = memcmp(&video->active_timings, &video->detected_timings,
+			    sizeof(video->active_timings));
+	video->active_timings = video->detected_timings;
+	aspeed_video_set_resolution(video);
+
+	if (is_res_chg) {
 		static const struct v4l2_event ev = {
 			.type = V4L2_EVENT_SOURCE_CHANGE,
 			.u.src_change.changes = V4L2_EVENT_SRC_CH_RESOLUTION,
-- 
2.17.1

